src/shutdown.rs

```diff
@@ -0,0 +1,56 @@
+use std::sync::Arc;
+use stream_cancel::{Trigger, Tripwire};
+
+/// When this struct goes out of scope it's a signal that its corresponding Source has completed
+/// executing and may be cleaned up.  It is the responsibility of each Source to ensure that this
+/// handle remains alive for the entire lifetime of the Source.  Source implementations must call
+/// take() on this at least once, and when multiple threads do work on behalf of a single Source
+/// each thread should own a clone of this and call take() on it at least once.
+#[derive(Clone)]
+pub struct SourceShutdownHandle {
+    _shutdown_complete: Arc<Trigger>,
+    taken: bool,
+}
+
+impl SourceShutdownHandle {
+    fn new(shutdown_complete: Trigger) -> Self {
+        Self {
+            _shutdown_complete: Arc::new(shutdown_complete),
+            taken: false,
+        }
+    }
+
+    /// Used to force moving this struct into a closure.
+    /// Every thread doing work on behalf of a Source must call this at least once.
+    pub fn take(&mut self) {
+        self.taken = true;
+    }
+}
+
+impl Drop for SourceShutdownHandle {
+    fn drop(&mut self) {
+        if !self.taken {
```

Comment:
I'd really like if this could be enforced at compile time rather than run time somehow.